# TFG
Hello, these are some of the scripts that I had to create in order to automate the use of Earl Grey, BRAKER3 (using the BRAKER2 tool) and GeMoMa. They serve as a presentation to my first 'professional' work.\
It is important to note that the code won't work if you just copy paste them because it was created using absolute routes (I thought it would be easier to run that way).
## Programs used
Here you can find the links to the articles that describe the use and making of the programs.\
Earl Grey:\
Baril, T., Galbraith, J., & Hayward, A. (2024). Earl Grey: A Fully Automated User-Friendly Transposable Element Annotation and Analysis Pipeline. Molecular Biology and Evolution, 41(4), msae068. https://doi.org/10.1093/molbev/msae068\
GeMoMa:\
Keilwagen, J., Wenk, M., Erickson, J. L., Schattat, M. H., Grau, J., & Hartung, F. (2016). Using intron position conservation for homology-based gene prediction. Nucleic Acids Research, 44(9), e89. https://doi.org/10.1093/nar/gkw092\
BRAKER3:\
Gabriel, L., Brůna, T., Hoff, K. J., Ebel, M., Lomsadze, A., Borodovsky, M., & Stanke, M. (2024). BRAKER3: Fully automated genome annotation using RNA-seq and protein evidence with GeneMark-ETP, AUGUSTUS and TSEBRA (p. 2023.06.10.544449). bioRxiv. https://doi.org/10.1101/2023.06.10.544449\
GAQET2:\
García-Carpintero, V. (2025). GAQET2 - Genome Annotation Quality Evaluation Tool. Published on the Internet; https://github.com/victorgcb1987/GAQET2 Retrieved 15 May 2025. Retrieved 28 of May 2025.![image](https://github.com/user-attachments/assets/e137cde9-d2c3-40a3-b2c5-d7d2e13f676a)\
